### PR TITLE
[GPU][DG2] Fix fusings_gpu/gemm_2in_act_scale_eltwise.basic/4

### PR DESCRIPTION
**Pull request by tuxedcat**
### Details:
 - Implement negative activation function
 - Remove workaround
 - Add testcases

### Tickets:
 - 67491

### CODE REVIEW

File: src/plugins/intel_gpu/tests/fusions/gemm_fusion_test.cpp [hunk 1]
```diff
@@ -376,19 +379,26 @@ TEST_P(gemm_2in_act_scale_eltwise, broadcast_eltwise) {
         reorder("reorder_bfyx", input_info("sum"), p.default_format, data_types::f32)
     );
     // Activation won't be fused because onednn doesn't support negative activation
-    if (engine.get_device_info().supports_immad && !p.kernel_name.empty())
+    if (engine.get_device_info().supports_immad) {
         p.expected_fused_primitives += 2;
+        if (!p.kernel_name.empty())
+            p.expected_not_fused_primitives += 2;
+    }
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/fusions/gemm_fusion_test.cpp:386]: we can support `negative` activation by mapping it into `eltwise_linear`

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/tests/fusions/gemm_fusion_test.cpp (hunk 1 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/fusions/gemm_fusion_test.cpp]: Done.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 2]
```diff
@@ -485,6 +485,9 @@ dnnl::algorithm convert_activation_func(cldnn::activation_func func) {
         case cldnn::activation_func::pow: return dnnl::algorithm::eltwise_pow;
         case cldnn::activation_func::sqrt: return dnnl::algorithm::eltwise_sqrt;
         case cldnn::activation_func::hard_sigmoid: return dnnl::algorithm::eltwise_hardsigmoid;
+        case cldnn::activation_func::square: return dnnl::algorithm::eltwise_square;
+        case cldnn::activation_func::hsigmoid: return dnnl::algorithm::undef;
+        case cldnn::activation_func::negative: return dnnl::algorithm::undef;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:490]: Please add a comment what "undef" means here.

File: src/plugins/intel_gpu/src/graph/program_node.cpp [hunk 3]
```diff
@@ -1290,20 +1301,10 @@ void program_node::init_onednn_primitive_attributes() {
...
-            update_onednn_post_op_list(onednn_post_op_type::eltwise_clip, empty_mem);
-            update_onednn_post_op_list(onednn_post_op_type::eltwise_linear, empty_mem);
-        } else {
-            auto params = get_fused_activations_params()[i];
-            dnnl::algorithm alg = onednn::convert_activation_func(activation_type);
-            post_ops.append_eltwise(1.0f, alg, params.a, params.b);
-            update_onednn_post_op_list(onednn_post_op_type::eltwise_act, empty_mem);
-        }
+        auto params = get_fused_activations_params()[i];
+        dnnl::algorithm alg = onednn::convert_activation_func(activation_type);
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/program_node.cpp:1305]: To be safe, we need to confirm that `alg != undef`. If arg is undef, we need to throw exception.

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/src/graph/program_node.cpp (hunk 3 above)
Comment (isanghao) [reply src/plugins/intel_gpu/src/graph/program_node.cpp:1306]: Please check this @tuxedcat

File: src/plugins/intel_gpu/tests/fusions/convolution_fusion_test.cpp [hunk 4]
```diff
@@ -2880,6 +2880,91 @@ INSTANTIATE_TEST_SUITE_P(fusings_gpu, conv_fp16_scale, ::testing::ValuesIn(std::
...
+        data("weights", get_mem(get_weights_layout(p))),
+        data("bias", get_mem(get_bias_layout(p))),
+        convolution("conv_prim", input_info("input"), { "weights" }, { "bias" }, p.groups, p.stride, p.pad, p.dilation),
+        activation("activation", input_info("conv_prim"), activation_func::negative),
+        reorder("reorder_bfyx", input_info("activation"), p.default_format, data_types::f32)
+    );
+
+    tolerance=default_tolerance(data_types::i8);
+    execute(p);
+}
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/fusions/convolution_fusion_test.cpp:2959]: I think we can combine these 5 cases into single implementation with parameter of `activation_func`.

**Review by isanghao — CHANGES_REQUESTED** [review]

**Review by isanghao — COMMENTED** [review]
Looks good to me. Let me check the performance.

**Comment by isanghao** [comment]
No perf impact ./check_performance.py --this ../log/2023-01-16--08-00.report --ref ../log/2023-01-12--08-46.report

Please rebase the PR @tuxedcat

File: src/plugins/intel_gpu/src/graph/program_node.cpp [hunk 5]
```diff
@@ -934,8 +934,22 @@ void program_node::init_onednn_primitive_attributes() {
                 post_ops.append_eltwise(dnnl::algorithm::eltwise_clip, 0.0f, 1.0f);
                 update_onednn_post_op_list(onednn_post_op_type::eltwise_linear, empty_mem);
                 update_onednn_post_op_list(onednn_post_op_type::eltwise_clip, empty_mem);
+            } else if (fused_desc->activation_function == cldnn::activation_func::hsigmoid) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/program_node.cpp:937]: I think we can use onednn `hard_sigmoid` for this case. Please see `eltwise_hardswish` example below.

File: src/plugins/intel_gpu/tests/fusions/convolution_fusion_test.cpp [hunk 6]
```diff
@@ -108,20 +129,21 @@ class ConvFusingTest : public BaseFusingTest<convolution_test_params> {
...
+    layout get_per_channel_layout(T_param& p) {
         return layout{ p.default_type, p.default_format, tensor{1, p.out_shape.feature[0], 1, 1} };
     }
 
-    layout get_prelu_slope_layout(convolution_test_params& p) {
+    layout get_prelu_slope_layout(T_param& p) {
         return layout{ p.default_type, p.input_format, tensor{1, p.out_shape.feature[0], p.out_shape.spatial[0], 1} };
     }
 };
+using ConvFusingTest=ConvFusingTestBase<convolution_test_params>;
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/fusions/convolution_fusion_test.cpp:146]: I'm afraid it makes debugging complicated. We already have several implementation of BaseFusingTest with different argument. What about just adding another implementation?

**Review by isanghao — CHANGES_REQUESTED** [review]
sorry for belated review. Please see my comment and ping me if you have any question.

**Review by isanghao — APPROVED** [review]

**Pull request merged**